openpype/pipeline/workfile/path_resolving.py

```diff
@@ -534,3 +541,50 @@ def create_workdir_extra_folders(
         fullpath = os.path.join(workdir, subfolder)
         if not os.path.exists(fullpath):
             os.makedirs(fullpath)
+
+
+def get_last_workfile_representation(
+    project_name: str,
+    asset_name: str,
+    task_name: str,
+    asset_id=None,
+    asset_doc: dict = None,
+    fields: Iterable[str] = None,
+) -> dict:
+    """Get last published workfile representation.
+
+    Args:
+        project_name(str): Project name.
+        asset_name(str): Asset/Shot name.
+        task_name (str): Task name.
+        asset_id (Optional[Union[str, ObjectId]]): Asset id.
+            subset name. Defaults to None.
+        asset_doc (Optional[dict]): Asset doc. Defaults to None.
+        fields (Optional[Iterable[str]]): Fields that should be returned.
+            Defaults to None.
+
+    Returns:
+        dict: Last workfile representation.
+    """
+    if not asset_doc:
+        asset_doc = get_asset_by_name(project_name, asset_name)
+
+    last_version = get_last_version_by_subset_name(
+        project_name,
+        get_subset_name(
+            "workfile",
+            "",
+            task_name,
+            asset_doc,
+            project_name=project_name,
+        ),
+        asset_id=asset_id,
+        asset_name=asset_name,
+        fields=["_id"],
+    )
+
+    return get_representations(
+        project_name,
+        version_ids=[last_version["_id"]],
+        fields=fields,
+    )[0]
```

Comment:
I'm not entirely convinced this is 100% reliable. There can even be cases where thumbnails also get integrated into the database and thus would also be a representation. See discussion [here](https://github.com/ynput/OpenPype/pull/5741#discussion_r1354572580). I believe Resolve actually publishes both a project workfile and an `.otio` file as part of the workfile family, which would introduce the same issue.

I wonder if the representation itself would have additional information that we can filter on to pick the right one that is the actual workfile we want.

Worst-case scenario we might maybe need to download each representation? Or need to filter by extension for a particular host, etc. but that'd of course introduce other issues?

Any ideas @iLLiCiTiT ?

Reply:
You should expect `workfile_extensions` as an argument. Also I don't think using `get_subset_name` should be used here, you are missing `host_name` and `variant`, which can be changed by user. You should probably get all subsets filter them by `"workfile"` family in data, find all latest versions, all their representations and based on matching extension with `workfile_extensions` the first could be used. It's also guessing, but more targeted.

Reply:
Ah ok, I wasn't sure if we were still maintaining old python versions.

Reply:
About using variants and extensions for filtering, I'm not sure the variant would change what which workfile is the right one, rather which variant of the same app will be launched, unless I'm missing something. I'm also not sure if it's necessary to check for extension, since the list of possible extensions depends on the host, so if I'm filtering by host I should get a valid workfile extension. Again, unless I'm missing something.

Reply:
@Sharkitty I must have missed something. Where are you filtering by host? :)

Reply:
@BigRoy In `get_subset_name`, I added it as an optional argument so it's taken in consideration. That being I'm not sure if I'll keep this method considering @iLLiCiTiT's comment

Reply:
It's not ensured the subset name would even include any details of the host - as such it doesn't ensure the subset you'll find is related to that host. Unfortunately that would not be a reliable filter.

Having said that - what we **should** be doing is storing in the **version data** when publishing which host or application was involved with the publish which could help in the future for decent filtering for something like this too. :)

Reply:
I refactored the whole method, what do you think about the new implementation?